### DISCUSSION TITLE
How to import project custom image related React components

### DISCUSSION THREAD

**Discussion by tiboue** [eclipse-sirius/sirius-components#3871] [Q&A]
Hello,

I am having some troubles importing some Sirius Web components related to a project custom images (such as [UploadImageModal.tsx](https://github.com/eclipse-sirius/sirius-web/blob/a688c6941ce49cde837174535c7c65276126fdd6/packages/sirius-web/frontend/sirius-web-application/src/views/project-settings/images/upload-image/UploadImageModal.tsx)). To my understanding, those should be accessible in the  sirius-web-application package with : 
`import { UploadImageModal } from '@eclipse-sirius/sirius-web-application';`

I unfortunately had no luck with version 2024.7.5 and got the following error: 

> Module '"@eclipse-sirius/sirius-web-application"' has no exported member 'UploadImageModal'

Any idea on how to access those components ?

Thanks.

**Comment by pcdavid** [comment]
`UploadImageModal` is not part of [the exported APIs](https://github.com/eclipse-sirius/sirius-web/blob/a688c6941ce49cde837174535c7c65276126fdd6/packages/sirius-web/frontend/sirius-web-application/src/index.ts) for that module, so currently there is no way to import them.

We try not export more than needed for concrete use cases, as exporting more stuff increases the surface API to be maintained.
What is your use case to need to import/reuse these?

**Reply by tiboue** [reply]
I would like to provide a way for users to display a custom image inside a node one. Using the already existing mechanism of custom images in the project settings seemed liked the logical way to do so.

For this, I am adding a Custom Widget directly accessible on the Nodes, with a "shortcut" to the modal rather than going through the project settings. While the shortcut is purely for usability sake, I still need to access the custom images of the project, but I had the same issue with [useProjectImages](https://github.com/eclipse-sirius/sirius-web/blob/26c980578f1df934e70f95323c5fb0592d1bfb7e/packages/sirius-web/frontend/sirius-web-application/src/views/project-settings/images/useProjectImages.ts).

**Reply by sbegaudeau** [reply]
> I would like to provide a way for users to display a custom image inside a node one. Using the already existing mechanism of custom images in the project settings seemed liked the logical way to do so.

The support for custom images has been created to let users leverage those custom images in their nodes so I don't understand why you would need something like that to make it works. Users can simply select the image in question as the shape to use and it just works.

**Reply by tiboue** [reply]
As far as I understand it, it is possible to specify an image (using an Image Style Description) for a type of node, but not for a specific instance of that Node. What I had in mind to solve this issue was to add a new Custom Node that displays images based on an aql expression (instead of using shape) to get the corresponding image.
The image specific to that instance of the Node would be set using a Custom Widget that allows adding new images and select one directly on the Node. It would be really helpful if this Custom Widget could reuse the UploadImageModal so that the end user can add new images without going back to project settings -> upload.
Exporting the UploadImageModal (and other project image related elements) would allow simpler widget definition that can be helpful in our case or other use case (decorators, etc.) while being certain that we are not breaking things by using a consistent API.

**Reply by pcdavid** [reply]
You can use `yalc` to develop your feature on a locally-modified verison of Sirius Components/Sirius Web with the additional exports you think you need, and propose a PR once you know the precise list.

**Reply by tiboue** [reply]
I have created a PR: https://github.com/eclipse-sirius/sirius-web/pull/4030. As mentionned, I only added the exports I requiered, you may want to consider adding additional ones.

**Comment by pcdavid** [comment]
The proposed PR been merged; closing.
